src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs

```diff
@@ -422,9 +450,17 @@ private void AddLexedTokenSlot()
             }
             else
             {
-                var tmp = new ArrayElement<SyntaxToken>[_lexedTokens.Length * 2];
-                Array.Copy(_lexedTokens, tmp, _lexedTokens.Length);
-                _lexedTokens = tmp;
+                var lexedTokens = _lexedTokens;
+
+                Array.Resize(ref _lexedTokens, _lexedTokens.Length * 2);
+
+                // Put lexedTokens back into the pool if it's correctly sized.
+                if (lexedTokens.Length == CachedTokenArraySize)
+                {
+                    // Fully clear out lexedTokens before releasing back to pool.
+                    Array.Clear(lexedTokens, 0, lexedTokens.Length);
+                    s_lexedTokensPool.Free(lexedTokens);
+                }
```

Comment:
Same question here (should we forget the tracking in the alternative).

Reply:
see other comment